controller/linodecluster_controller.go

```diff
@@ -18,45 +18,216 @@ package controller
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.0/pkg/reconcile
+
 func (r *LinodeClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(r.ReconcileTimeout))
+	defer cancel()
+
+	logger := ctrl.LoggerFrom(ctx).WithName("LinodeClusterReconciler").WithValues("name", req.NamespacedName.String())
+	LinodeCluster := &infrav1alpha1.LinodeCluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, LinodeCluster); err != nil {
+		logger.Info("Failed to fetch Linode cluster", "error", err.Error())
+
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	cluster, err := kutil.GetOwnerCluster(ctx, r.Client, LinodeCluster.ObjectMeta)
+	if err != nil {
+		logger.Info("Failed to get owner cluster", "error", err.Error())
+
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	} else if cluster == nil {
+		logger.Info("Machine Controller has not yet set OwnerRef, skipping reconciliation")
+
+		return ctrl.Result{}, nil
+	}
+	if annotations.IsPaused(cluster, LinodeCluster) {
+		logger.Info("LinodeCluster of linked Cluster is marked as paused. Won't reconcile")
+
+		return ctrl.Result{}, nil
+	}
+	// Create the cluster scope.
+	clusterScope, err := scope.NewClusterScope(
+		r.LinodeApiKey,
+		scope.ClusterScopeParams{
+			Client:        r.Client,
+			Cluster:       cluster,
+			LinodeCluster: LinodeCluster,
+		})
+	if err != nil {
+		logger.Info("Failed to create cluster scope", "error", err.Error())
+
+		return ctrl.Result{}, fmt.Errorf("failed to create cluster scope: %w", err)
+	}
+
+	return r.reconcile(ctx, clusterScope, logger)
+}
+
+func (r *LinodeClusterReconciler) reconcile(
+	ctx context.Context,
+	clusterScope *scope.ClusterScope,
+	logger logr.Logger,
+) (res ctrl.Result, reterr error) {
+	res = ctrl.Result{}
+
+	clusterScope.LinodeCluster.Status.Ready = false
+	clusterScope.LinodeCluster.Status.FailureReason = nil
+	clusterScope.LinodeCluster.Status.FailureMessage = util.Pointer("")
+
+	// Always close the scope when exiting this function so we can persist any GCPMachine changes.
+	defer func() {
+		if err := clusterScope.Close(); utilerrors.FilterOut(err, apierrors.IsNotFound) != nil && reterr == nil {
+			logger.Error(err, "failed to patch LinodeCluster")
+			reterr = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !clusterScope.LinodeCluster.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, r.reconcileDelete(ctx, logger, clusterScope)
+	}
+
+	controllerutil.AddFinalizer(clusterScope.LinodeCluster, infrav1alpha1.GroupVersion.String())
+	// Create
+	if clusterScope.LinodeCluster.Spec.ControlPlaneEndpoint.Host == "" {
+		if err := r.reconcileCreate(ctx, clusterScope, logger); err != nil {
```

Comment:
```
return ctrl.Result{}, r.reconcileCreate(ctx, clusterScope, logger)
```
for consistency with above `reconcileDelete` (also swap the argument order for consistency between `reconcileCreate` and `reconcileDelete`)

Reply:
in this case we don't want to return if there is an error so we can set the status and conditions I originally kept those outside the scope of the create function so we can add update functions later but I can combine them if we think that would be better